modin/core/storage_formats/base/query_compiler.py

```diff
@@ -3089,7 +3089,145 @@ def get_axis(self, axis):
         """
         return self.index if axis == 0 else self.columns
 
-    def take_2d(self, index=None, columns=None):
+    def take_2d_labels(
+        self,
+        index,
+        columns,
+    ):
+        """
+        Take the given labels.
+
+        Parameters
+        ----------
+        index : slice, scalar, list-like, or BaseQueryCompiler
+            Labels of rows to grab.
+        columns : slice, scalar, list-like, or BaseQueryCompiler
+            Labels of columns to grab.
+
+        Returns
+        -------
+        BaseQueryCompiler
+            Subset of this QueryCompiler.
+        """
+        row_lookup, col_lookup = self.get_positions_from_labels(index, columns)
+        if isinstance(row_lookup, slice):
+            ErrorMessage.catch_bugs_and_request_email(
+                failure_condition=row_lookup != slice(None),
+                extra_log=f"Only None-slices are acceptable as a slice argument in masking, got: {row_lookup}",
+            )
+            row_lookup = None
+        if isinstance(col_lookup, slice):
+            ErrorMessage.catch_bugs_and_request_email(
+                failure_condition=col_lookup != slice(None),
+                extra_log=f"Only None-slices are acceptable as a slice argument in masking, got: {col_lookup}",
+            )
+            col_lookup = None
+        return self.take_2d_positional(row_lookup, col_lookup)
+
+    def get_positions_from_labels(self, row_loc, col_loc):
+        """
+        Compute index and column positions from their respective locators.
+
+        Inputs to this method are arguments the the pandas user could pass to loc.
+        This function will compute the corresponding index and column positions
+        that the user could equivalently pass to iloc.
+
+        Parameters
+        ----------
+        row_loc : scalar, slice, list, array or tuple
+            Row locator.
+        col_loc : scalar, slice, list, array or tuple
+            Columns locator.
```

Comment:
judging by the implementation, this function may also take pandas.Index, MultiIndex, RangeIndex, and all of the variety of non-standardized objects. Do we really want the logic that's supposed to parse all of this zoo inside the query compiler? I would expect to have some kind of `preprocess_labels(...)` function at the `modin.pandas.indexing` that converts a comprehensive variety of objects that pandas can consume into something simple and standardized and THEN pass it to the query compiler.